zipline/algorithm.py

```diff
@@ -457,30 +433,40 @@ def _create_generator(self, sim_params, source_filter=None):
         processed by the zipline, and False for those that should be
         skipped.
         """
-
-        if not self.initialized:
-            self.initialize(*self.initialize_args, **self.initialize_kwargs)
-            self.initialized = True
+        if sim_params is not None:
+            self.sim_params = sim_params
 
         if self.perf_tracker is None:
+            # Build a perf_tracker
+            self.perf_tracker = PerformanceTracker(
+                sim_params=self.sim_params,
+                env=self.trading_environment,
+                data_portal=self.data_portal)
+            # Set the dt initially to the period start by forcing it to change
+            self.on_dt_changed(self.sim_params.period_start)
+
             # HACK: When running with the `run` method, we set perf_tracker to
             # None so that it will be overwritten here.
             self.perf_tracker = PerformanceTracker(
-                sim_params=sim_params, env=self.trading_environment
+                sim_params=sim_params, env=self.trading_environment,
+                data_portal=self.data_portal
             )
 
-        self.portfolio_needs_update = True
-        self.account_needs_update = True
-        self.performance_needs_update = True
-
-        self.data_gen = self._create_data_generator(source_filter, sim_params)
+        if not self.initialized:
+            self.initialize(*self.initialize_args, **self.initialize_kwargs)
+            self.initialized = True
 
-        self.trading_client = AlgorithmSimulator(self, sim_params)
+        self.ensure_clock()
 
-        transact_method = transact_partial(self.slippage, self.commission)
-        self.set_transact(transact_method)
+        self.trading_client = AlgorithmSimulator(
+            self,
+            sim_params,
+            self.data_portal,
+            self.clock,
```

Comment:
Agreed.  Removed it as an instance attribute of the algorithm, and moved clock creation to `Algorithm._create_clock` (which others can override if needed).